lib/index.js

```diff
@@ -0,0 +1,281 @@
+ */
+function fetchMessage(url) {
+  return axios
+    .get(url, {responseType: 'json', headers: {accept: 'application/json'}})
+    .then(res => res.data);
+}
+
+/**
+ * @param keyId {string}
+ * @returns {string|null}
+ */
+function didFromKeyId(keyId) {
+  if(!keyId) {return null;}
+
+  return keyId.split('#')[0]; // Drop hash fragment
+}
+
+/**
+ * @param options {object} Options hashmap (see docstring for `getDidDoc()`)
+ * @param options.did {string}
+ * @param options.location {string} Typically 'local' (when decrypting)
+ *   or 'all' (when encrypting).
+ *
+ * @param [keyId] {string} Optional key id, if known
+ *
+ * @throws {Error} If unable to fetch a local DID Document
+ *
+ * @returns {{publicKey: Buffer, secretKey: Buffer}} KeyPair (curve25519
+ *   encryption keys)
+ */
+async function cipherKeysForDid(options, keyId) {
+  const doc = await getDidDoc(options);
+
+  if(!doc) {
+    throw new Error(`Could not fetch DID Document for ${options.did}`);
+  }
+
+  return cipherKeysFor(doc, keyId);
+}
+
+/**
+ * @param doc {Document} DID Document
+ * @param [keyId] {string} Optional key id, if known
+ *
+ * @returns {{publicKey: Buffer, secretKey: Buffer}} KeyPair (curve25519
+ *   encryption keys)
+ */
+function cipherKeysFor(doc, keyId = null) {
+  const keys = authKeysFor(doc, keyId);
+
+  const {
+    publicKey: signKeyPublic,
+    secretKey: signKeySecret
+  } = keys;
+  keyId = keyId || keys.keyId;
+
+  // Convert DID ed25519 signing authentication keys to curve25519 keys
+  const cipherKeyPublic = new Buffer(sodium.crypto_box_PUBLICKEYBYTES);
+  sodium.crypto_sign_ed25519_pk_to_curve25519(cipherKeyPublic, signKeyPublic);
+
+  let cipherKeySecret;
+  if(signKeySecret) {
+    cipherKeySecret = new Buffer(sodium.crypto_box_SECRETKEYBYTES);
+    sodium.crypto_sign_ed25519_sk_to_curve25519(cipherKeySecret, signKeySecret);
+  }
+
+  return {
+    keyId,
+    publicKey: cipherKeyPublic,
+    secretKey: cipherKeySecret
+  };
+}
+
+/**
+ * @param doc {Document} DID Document
+ * @param [keyId] {string} Optional key id, if known
+ *
+ * @returns {{publicKey: Buffer, secretKey: Buffer}} KeyPair (ed25519)
+ */
+function authKeysFor(doc, keyId) {
+  const authKey = doc.authentication[0].publicKey[0]; // ed25519 signing key
+
+  const publicKey = bs58.decode(authKey.publicKeyBase58);
+  let secretKey;
+
+  if(authKey.privateKey) {
+    secretKey = bs58.decode(authKey.privateKey.privateKeyBase58);
+  }
+
+  return {
+    publicKey,
+    secretKey,
+    keyId: authKey.id
+  };
+}
+
+/**
+ * Resolves with a DID Document fetch result or null if not found.
+ *
+ * @param options {object} Options hashmap
+ * @param options.did {string}
+ * @param options.mode {string}
+ * @param options.location {string}
+ * @param [options.client]
+ *
+ * @returns {Promise<Document|null>}
+ */
+function getDidDoc(options) {
+  const client = options.client || veres;
+
+  return new Promise((resolve, reject) => {
+    client.info({
```

Comment:
For a demo this is fine for now, but will want to use some other API like did-io soon.  That API was not meant to be called like this.  I think it could call the callback many times with certain options.

Reply:
Yeah, agreed, will be changed as soon as did-io refactors are done.